File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    class Pair{
        int idx;
        int height;

        Pair(int idx, int height){
            this.idx = idx;
            this.height = height;
        }
    }

    public int largestRectangleArea(int[] heights) {
        Stack<Pair> stack = new Stack<>();
        int max = 0;

        for(int i=0; i<heights.length; i++){
            int startIdx = i;

            while(!stack.isEmpty() && stack.peek().height > heights[i]){
                Pair top = stack.pop();
                max = Math.max(max, top.height * (i - top.idx));
                startIdx = top.idx;
            }

            stack.push(new Pair(startIdx, heights[i]));
        }

        while(!stack.isEmpty()){
            Pair top = stack.pop();
            max = Math.max(max, top.height * (heights.length - top.idx));
        }

        return max;
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution(object):
    def largestRectangleArea(self, heights):
        heights.append(0)  # Sentinel to flush the stack
        stack = []
        max_area = 0

        for i, h in enumerate(heights):
            while stack and h < heights[stack[-1]]:
                top = stack.pop()
                height = heights[top]
                width = i if not stack else i - stack[-1] - 1
                max_area = max(max_area, height * width)
            stack.append(i)

        return max_area

